### PR TITLE
WIP: KP-6483 Depth-first download procedure

### PR DESCRIPTION
**Pull request by helhotti**

### CODE REVIEW

File: pipeline/dags/parallel_download.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+        # This is how it would preferrably be done:
+        # expand() only accepts lists or dicts, so dc_identifiers need to be
+        # converted to a list. With the current implementation, this would
+        # be an issue for large sets, but we have an upcoming ticket to
+        # download huge sets in batches anyway, which should hopefully fix
+        # this issue.
+        # dc_identifiers = list(api.dc_identifiers(set_id))
+
+        @task(
+            task_id=f"download_binding", task_group=download, trigger_rule=trigger_rule
```
Comment (aajarven) [inline pipeline/dags/parallel_download.py:93]: This f-stringness looks unnecessary

File: pipeline/dags/parallel_download.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+        Check if API is responding and download can begin. If not, cancel pipeline.
+        """
+        api_ok = HttpSensor(
+            task_id="http_sensor", http_conn_id="nlf_http_conn", endpoint="/"
+        ).poke(context={})
+
+        if api_ok == True:
+            return "begin_download"
+        elif api_ok == False:
+            return "cancel_pipeline"
```
Comment (aajarven) [inline pipeline/dags/parallel_download.py:161]: Could we just do `if api_ok:` accompanied by `else`? Or if it's important that False is different from None, then `if api_ok is True` + `elif api_ok is False` is more pythonic.

**Review by aajarven — CHANGES_REQUESTED** [review]
The dag structure makes sense and looks like it should work as we want. I had a couple of style improvement comments, and a possible trigger rule improvement (if it works as I think it does), but otherwise we are good to go.

If there's time, I guess now that we are nearing what might be the final structure it would be a good point to see if we have stuff that could be cleaned up, like unused dags / operators or duplicate code that could be refactored into reusable parts or whatnot.

File: pipeline/dags/parallel_download.py (hunk 1 above)
Comment (helhotti) [reply pipeline/dags/parallel_download.py:93]: It does indeed, this is leftover from when I was hopeful that we could somehow give a custom name for each mapped task with `expand()`. Turns out we cannot without creating the tasks in a loop.

File: pipeline/dags/parallel_download.py (hunk 2 above)
Comment (helhotti) [reply pipeline/dags/parallel_download.py:161]: I was first weary of cases where the poking function may decide to not return anything, but now that I think about it we probably want to cancel the pipeline in those cases as well (and it shouldn't ever happen anyway), so `if api_ok:` and `else` should work just fine.

File: pipeline/dags/parallel_download.py [hunk 3]
```diff
@@ -157,15 +175,24 @@ def download_binding(dc_identifier):
...
-        )
         if not downloads:
-            check_api_availability >> download_tg
+            download_tg = download_set(
+                dag=dag,
+                set_id=set_id,
+                api=api,
+                ssh_conn_id=SSH_CONN_ID,
+                base_path=BASE_PATH,
+                trigger_rule="all_success",
```
[the remark this answers is not in the document]
Comment (helhotti) [reply pipeline/dags/parallel_download.py:185]: You're right, we don't!

File: pipeline/dags/parallel_download.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+    CreateConnectionOperator,
+)
+
+HTTPConnection.debuglevel = 1
+
+logging.basicConfig(filename="/home/ubuntu/airflow/logs/request_logs/request_logs.log")
+logging.getLogger().setLevel(logging.DEBUG)
+requests_log = logging.getLogger("requests.packages.urllib3")
+requests_log.setLevel(logging.DEBUG)
+requests_log.propagate = True
```
[the remark this answers is not in the document]
Comment (helhotti) [reply pipeline/dags/parallel_download.py:33]: Probably not, I'm now somewhat aware of when and why API calls are made. If we need it again after the rate limit issues are hopefully solved, we can put it back.

**Comment by helhotti** [comment]
DAG is now refactored to address Anni's review.

**Review by aajarven — APPROVED** [review]
Yay, looks good!

**Pull request merged**